cpp/src/arrow/datum.h

```diff
@@ -43,14 +43,18 @@ class Table;
 /// \class Datum
 /// \brief Variant type for various Arrow C++ data structures
 struct ARROW_EXPORT Datum {
+  /// \brief The kind of datum stored
   enum Kind { NONE, SCALAR, ARRAY, CHUNKED_ARRAY, RECORD_BATCH, TABLE };
 
+  /// \brief A placeholder type to represent empty datum
   struct Empty {};
 
-  // Datums variants may have a length. This special value indicate that the
-  // current variant does not have a length.
+  /// \brief Datums variants may have a length. This special value indicate that the
+  /// current variant does not have a length.
   static constexpr int64_t kUnknownLength = -1;
 
+  /// \brief Storage of the actual datum. Note: For arrays, it stores ArrayData instead of
+  /// Array for easier data processing
```

Comment:
Let's keep the brief summary brief and move the rest to the longer description section?
```suggestion
  /// \brief Storage of the actual datum.
  ///
  /// Note: For arrays, ArrayData is stored instead of Array for easier processing
```